tests/dogfood.rs

```diff
@@ -25,43 +25,206 @@ fn dogfood_clippy() {
     }
 }
 
-#[test]
-#[ignore]
-#[cfg(feature = "internal")]
-fn run_metadata_collection_lint() {
-    use std::fs::File;
-    use std::time::SystemTime;
-
-    // Setup for validation
-    let metadata_output_path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("util/gh-pages/lints.json");
-    let start_time = SystemTime::now();
-
-    // Run collection as is
-    std::env::set_var("ENABLE_METADATA_COLLECTION", "1");
-    run_clippy_for_package("clippy_lints", &["-A", "unfulfilled_lint_expectations"]);
-
-    // Check if cargo caching got in the way
-    if let Ok(file) = File::open(metadata_output_path) {
-        if let Ok(metadata) = file.metadata() {
-            if let Ok(last_modification) = metadata.modified() {
-                if last_modification > start_time {
-                    // The output file has been modified. Most likely by a hungry
-                    // metadata collection monster. So We'll return.
-                    return;
+#[cfg(all(test, feature = "internal"))]
+mod internal {
+    use super::run_clippy_for_package;
+    use if_chain::if_chain;
+    use serde::Deserialize;
+    use std::collections::HashMap;
+    use std::ffi::OsStr;
+    use std::fs::{read_to_string, File};
+    use std::io::BufReader;
+    use std::path::PathBuf;
+    use walkdir::WalkDir;
+
+    static UI_TEST_PATH: std::lazy::SyncLazy<PathBuf> =
+        std::lazy::SyncLazy::new(|| PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/ui"));
+
+    static METADATA_OUTPUT_PATH: std::lazy::SyncLazy<PathBuf> =
+        std::lazy::SyncLazy::new(|| PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("util/gh-pages/lints.json"));
+
+    #[derive(Debug, Deserialize)]
+    pub struct LintMetadata {
+        id: String,
+        applicability: Option<ApplicabilityInfo>,
+    }
+
+    #[derive(Debug, Deserialize)]
+    struct ApplicabilityInfo {
+        applicability: Option<String>,
+    }
+
+    const MACHINE_APPLICABLE: &str = "MachineApplicable";
+
+    const KNOWN_EXCEPTIONS: [&str; 33] = [
```

Comment:
Here you can use a reference to not hard-code the array length :upside_down_face: 
```suggestion
    const KNOWN_EXCEPTIONS: &[&str] = &[
```